File: JavaScript/block_sort.js
function blockSort(arr, blockSize) {
    if (blockSize <= 0) {
        throw new Error("Block size must be greater than 0");
    }

    const n = arr.length;
    const sortedArr = [];

    for (let i = 0; i < n; i += blockSize) {
        const block = arr.slice(i, i + blockSize);
        block.sort((a, b) => a - b);
        sortedArr.push(...block);
    }

    return mergeSortedBlocks(sortedArr, blockSize);
}

function mergeSortedBlocks(arr, blockSize) {
    const n = arr.length;
    const result = [];
    const numBlocks = Math.ceil(n / blockSize);
    const pointers = Array(numBlocks).fill(0);

    while (result.length < n) {
        let minIndex = -1;
        let minValue = Infinity;

        for (let i = 0; i < numBlocks; i++) {
            const blockStart = i * blockSize;
            const blockEnd = Math.min(blockStart + blockSize, n);
            const pointer = pointers[i];

            if (pointer < blockEnd - blockStart) {
                const value = arr[blockStart + pointer];
                if (value < minValue) {
                    minValue = value;
                    minIndex = i;
                }
            }
        }

        if (minIndex !== -1) {
            result.push(minValue);
            pointers[minIndex]++;
        }
    }

    return result;
}

const arr = [5, 2, 9, 1, 5, 6, 0];
const blockSize = 2;
console.log(blockSort(arr, blockSize));

File: Kotlin/blockSort.kt
fun blockSort(arr: IntArray, blockSize: Int): IntArray {
    if (arr.isEmpty()) return arr

    val n = arr.size
    val sortedArr = arr.copyOf()
    val blocks = (n + blockSize - 1) / blockSize

    for (i in 0 until blocks) {
        val start = i * blockSize
        val end = minOf(start + blockSize, n)
        sortedArr.sort(start, end)
    }

    val result = IntArray(n)
    var index = 0
    for (i in 0 until blocks) {
        val start = i * blockSize
        val end = minOf(start + blockSize, n)
        for (j in start until end) {
            result[index++] = sortedArr[j]
        }
    }
    result.sort()

    return result
}

fun main() {
    val arr = intArrayOf(5, 2, 9, 1, 5, 6, 0)
    val blockSize = 2
    val sortedArr = blockSort(arr, blockSize)
    println(sortedArr.joinToString(", "))
}